R/dash.R

```diff
@@ -505,55 +701,280 @@ Dash <- R6::R6Class(
     
+    # ------------------------------------------------------------------------
+    # return asset URLs
+    # ------------------------------------------------------------------------
+    get_asset_url = function(asset_path, prefix = self$config$requests_pathname_prefix) {
+      app_root_path <- Sys.getenv("DASH_APP_PATH")
+      
+      if (app_root_path == "" && getAppPath() != FALSE) {
+        # app loaded via source(), root path is known
+        app_root_path <- dirname(private$app_root_path)
+      } else if (getAppPath() == FALSE) {
+        # app not loaded via source(), env var not set, no reliable way to ascertain root path
+        warning("application not started via source(), and DASH_APP_PATH environment variable is undefined. get_asset_url returns NULL since root path cannot be reliably identified.")
+        return(NULL)
+      }
+      
+      asset <- lapply(private$asset_map, 
+                      function(x) {
+                        # asset_path should be prepended with the full app root & assets path
+                        # if leading slash(es) present in asset_path, remove them before
+                        # assembling full asset path
+                        asset_path <- file.path(app_root_path,
+                                                private$assets_folder, 
+                                                sub(pattern="^/+",
+                                                    replacement="",
+                                                    asset_path))
+                        return(names(x[x == asset_path]))
+                      }
+      )
+      asset <- unlist(asset, use.names = FALSE)
+      
+      if (length(asset) == 0)
+        stop(sprintf("the asset path '%s' is not valid; please verify that this path exists within the '%s' directory.",
+                     asset_path,
+                     private$assets_folder))
+      
+      # strip multiple slashes if present, since we'll
+      # introduce one when we concatenate the prefix and
+      # asset path & prepend the asset name with route prefix
+      return(gsub(pattern="/+",
+                  replacement="/",
+                  paste(prefix, 
+                        private$assets_url_path, 
+                        asset, 
+                        sep="/")))
+    },
+    
+    # ------------------------------------------------------------------------
+    # specify a custom index string
+    # ------------------------------------------------------------------------
+    index_string = function(string) {
+      if (!grepl("react-entry-point", string)) stop("React entry point div must be defined within the index string.")
+      if (!grepl("_dash-config", string)) stop("Dash config script has not been defined within the index string.")
+      if (!grepl("scripts_tags", string)) stop("{scripts_tags} must be included to render the Dash app and any JavaScript files in /assets/.")
+      private$custom_index = string
+    },
```

Comment:
On the Python side, it looks like the analogous code block is

```
    @index_string.setter
    def index_string(self, value):
        checks = (
            (_re_index_entry.search(value), "app_entry"),
            (_re_index_config.search(value), "config"),
            (_re_index_scripts.search(value), "scripts"),
        )
        missing = [missing for check, missing in checks if not check]
        if missing:
            raise exceptions.InvalidIndexException(
                "Did you forget to include {} in your index string ?".format(
                    ", ".join("{%" + x + "%}" for x in missing)
                )
            )
        self._index_string = value
```

We can leave the (internal) variable names as they are for now, but the interface should probably match the Python side, so that we're looking for `app_entry`, `config`, and `scripts`. 

It would also be good if we could throw an error message which includes all the elements that are missing, as Python also does -- that way if 2/3 or 3/3 need fixing, the user will be notified of all of them. (Since here if both `_dash-config` and `react-entry-point` are not defined, you'll only know that the latter is wrong, then get prompted again that `_dash-config` must be defined too.)

Here's an existing example that does something similar (in terms of returning an message with a variable number of errors):

https://github.com/plotly/dashR/blob/f81fcb5332eb66a0933d3fd5b1a404279fb3c459/R/utils.R#L290-L300

Reply:
Addressed in 1138e4. Might be worth taking another look at again though, just to make sure the index key changes are the ones we want to stick with moving forward, this did involve changes to the default index as well.

Also made another change in 17a61d, this moves the `react-entry-point` div to the {app_entry} key itself. From a user standpoint it's better because it makes the custom index less complex and easier to manipulate without breaking the essential components.